Make pause overlay close with pause gameplay action

Joehuu opened the pull request:
- Noticed in https://github.com/ppy/osu/pull/22000#issuecomment-1369041351

This behavior matches stable (fail overlay does nothing with middle mouse, so the action is localized to `PauseOverlay`).

Joehuu committed:
Make pause overlay close with pause gameplay action

peppy commented:
This is a kind of change I'd expect to have test coverage (because we've regularly regressed shortcuts by handling them at a high or lower level).

peppy reviewed (changes requested):
Looks fine, but requires test coverage.

frenzibyte commented:
In addition, the setting should probably be renamed to "Pause / resume gameplay".

Joehuu committed:
Fix manual input manager test scenes not matching game input hierarchy

Fix popover using on key down

Fix popover not expiring when using global action

Joehuu committed:
Refactor pause test to actually use back action keybinding

Joehuu committed:
Refactor pause test to actually use quick exit action keybinding

Joehuu committed:
Add basic toggle pause tests

Joehuu committed:
Rename pause gameplay keybind to reflect other behavior

peppy commented on osu.Game/Graphics/UserInterfaceV2/OsuPopover.cs:
```diff
@@ -58,6 +60,14 @@ protected override void PopOut()
             this.FadeOut(fade_duration, Easing.OutQuint);
         }
 
+        protected override bool OnKeyDown(KeyDownEvent e)
```
Can you explain why these changes were required? I can't immediately see why from your commit history, and it looks unrelated to the location you're actually changing things (gameplay doesn't use any popovers to my knowledge).

Joehuu replied:
[776b60f](https://github.com/ppy/osu/pull/22079/commits/776b60f3b3b133ecd41bc284e727e333c670f949) made global actions handle first in `ManualInputManagerTestScene`s. It made a popover test fail, and was fixed alongside.

The test failed because the popover never expired due to the global action using `Hide()` instead of `this.HidePopover()` from the `OnKeyDown` implementation. I fixed that but also removed the handling from `OnKeyDown` similar to `FocusedTextBox`.

peppy commented on osu.Game/Tests/Visual/OsuManualInputManagerTestScene.cs:
```diff
@@ -46,21 +46,23 @@ protected OsuManualInputManagerTestScene()
+
             if (DisplayCursorForManualInput)
             {
                 var cursorDisplay = new GlobalCursorDisplay { RelativeSizeAxes = Axes.Both };
 
-                cursorDisplay.Add(new OsuTooltipContainer(cursorDisplay.MenuCursor)
+                cursorDisplay.Add(content = new OsuTooltipContainer(cursorDisplay.MenuCursor)
                 {
                     RelativeSizeAxes = Axes.Both,
-                    Child = mainContent
                 });
 
-                mainContent = cursorDisplay;
+                inputContainer.Add(cursorDisplay);
+                mainContent = inputContainer;
             }
 
             if (CreateNestedActionContainer)
-                mainContent = new GlobalActionContainer(null).WithChild(mainContent);
+                inputContainer.Add(new GlobalActionContainer(null));
```
I can't understand what this code is doing anymore. Probably best to restructure into a single hierarchy if possible.

Joehuu replied:
Have simplified it. Not sure why I made another container.

peppy commented on osu.Game/Screens/Play/PauseOverlay.cs:
```diff
@@ -56,5 +58,17 @@ protected override void PopOut()
+            switch (e.Action)
+            {
+                case GlobalAction.PauseGameplay:
+                    BackAction.Invoke();
```
Bit unfortunate that this is calling `BackAction`.

Joehuu committed:
Simplify container logic

Joehuu committed:
Use `TriggerClick()` instead of calling `BackButton` action for pause gameplay

peppy commented on osu.Game/Screens/Play/PauseOverlay.cs:
```diff
@@ -64,7 +64,7 @@ public override bool OnPressed(KeyBindingPressEvent<GlobalAction> e)
             {
                 case GlobalAction.PauseGameplay:
-                    BackAction.Invoke();
+                    InternalButtons.Children.First().TriggerClick();
```
```suggestion
                    InternalButtons.First().TriggerClick();
```

I think this should work (and read better).

Joehuu committed:
Simplify `TriggerClick`s in gameplay menu overlays

peppy committed:
Merge branch 'master' into fix-pause-gameplay-action-not-closing
